pashage.1
.Dd November 2, 2025
.Dt PASHAGE 1
.Os
.Sh NAME
.Nm pashage
.Nd simple and portable password store for the Unix shell
.Sh SYNOPSIS
.Nm
.Op Ar COMMAND
.Op Ar OPTIONS
.Op Ar ARGS
.Sh DESCRIPTION
.Nm
is a
.Em password manager ,
which means it manages a database of encrypted secrets, including encrypting
externally-provided new secrets, generating and encrypting random strings, and
decrypting and displaying stored secrets.
.Pp
It aims to be simple and composable, but its reliance on Unix philosophy
and customs might make steep learning curve for users outside of this
culture.
.Pp
The
.Nm
utility provides commands for query and management of the password store.
When no command is specified,
.Cm list
or
.Cm show
is implicitly assumed.
.Pp
The database is optionally versioned using
.Xr git 1
to help with history audit and synchronization.
It should be noted that this prevents re-encryption from erasing old
cyphertext, leaving the secret vulnerable to compromised encryption keys.
.Pp
The cryptography is done by
.Xr age 1
external command.
It decrypts using the
.Em identity
file given in the environment, and crypts using a list of
.Em recipients
per subfolder, defaulting to the parent
.Em recipient
list or the
.Em identity .
.Sh COMMANDS
.Ss copy
.Nm
.Cm copy
.Op Fl e,--reencrypt | Fl i,--interactive | Fl k,--keep
.Op Fl f,--force
.Ar old-path ... new-path
.Pp
This subcommand copies secrets and recursively copies subfolders,
using the same positional argument scheme as
.Xr cp 1 .
By default it asks before overwriting an existing secret and it re-encrypts
the secret when the destination has a different
.Em recipient
list.
.Pp
The options are as follows:
.Bl -tag -compact -width \-i,--interactive
.It Fl e,--reencrypt
always re-encrypt secrets
.It Fl f,--force
overwrite existing secrets without asking
.It Fl i,--interactive
asks whether to re-encrypt or not for each secret
.It Fl k,--keep
never re-encrypt secrets
.El
.Ss delete
.Nm
.Cm delete
.Op Fl r,--recursive
.Op Fl f,--force
.Ar pass-name
.Ar ...
.Pp
This subcommand deletes secrets from the database.
By default it skips subfolders and asks for confirmation for each secret.
.Pp
The options are as follows:
.Bl -tag -compact -width \-r,--recursive
.It Fl f,--force
delete without asking for confirmation
.It Fl r,--recursive
recursively delete all secrets in given subfolders
.El
.Ss edit
.Nm
.Cm edit
.Ar pass-name
.Ar ...
.Pp
This subcommand starts an interactive editor to update the secrets.
.Ss find
.Nm
.Cm find
.Op Fl r,--raw
.Op Ar GREP_OPTIONS
.Ar regex
.Pp
This subcommand lists as a tree the secrets whose name match the given
regular expression, using the corresponding
.Xr grep 1
options.
.Pp
The options are as follows:
.Bl -tag -compact -width \-r,--raw
.It Fl r,--raw
display the results as a raw list of secrets, rather than a tree
.El
.Ss generate
.Nm
.Cm generate
.Op Fl n,--no-symbols
.Op Fl c,--clip | Fl q,--qrcode
.Op Fl i,--in-place | Fl f,--force
.Op Fl m,--multiline
.Op Fl t,--try
.Ar pass-name
.Op Ar pass-length Op Ar character-set
.Pp
This subcommand generates a new secret from
.Pa /dev/urandom ,
stores it in the database, and by default displays it on the standard output
and asks
for confirmation before overwriting an existing secret.
.Pp
The options are as follows:
.Bl -tag -compact -width \-n,--no-symbols
.It Fl c,--clip
paste the secret into the clipboard instead of using the standard output
.It Fl f,--force
replace existing secrets without asking
.It Fl i,--in-place
when the secret already exists, replace only its first line and re-use the
following lines
.It Fl m,--multiline
read lines from standard input to append after the generated data into the secret
file
.It Fl n,--no-symbols
generate a secret using only alphanumeric characters
.It Fl q,--qrcode
display the secret as a QR-code instead of using the standard output
.It Fl t,--try
display the secret and ask for confirmation before storing it into the database
.El
.Ss git
.Nm
.Cm git git-command-args ...
.Pp
This subcommand invokes
.Xr git 1
in the database repository.
Only
.Cm git init
and
.Cm git clone
are accepted when there is no underlying repository.
.Ss gitconfig
.Nm
.Cm gitconfig
.Pp
This subcommand configures the underlying repository to automatically
decrypt secrets to display differences.
.Ss grep
.Nm
.Cm grep
.Op Ar GREP_OPTIONS
.Ar search-regex
.Pp
This subcommand successively decrypts all the secrets in the store and
filter them through
.Xr grep 1
using the given options, and outputs all the matching lines and the
corresponding secret.
.Ss help
.Nm
.Cm help
.Op Ar subcommand ...
.Pp
This subcommand displays on the standard output the help text for the
given subcommands.
.Pp
Without arguments, this subcommand displays on the standard output the
version and help text, including all subcommands and flags and a brief
description.
.Ss init
.Nm
.Cm init
.Op Fl i,--interactive | Fl k,--keep
.Op Fl p,--path Ar subfolder
.Ar age-recipient
.Ar ...
.Pp
This subcommand initializes an age
.Em recipient
list, by default of the root of the password store, and re-encrypts all the
affected secrets.
When the
.Ar age-recipient
list is a single empty string, the
.Em recipient
list is instead removed, falling back to a parent
.Em recipient
list or ultimately to the age
.Em identity .
.Pp
The options are as follows:
.Bl -tag -compact -width \-i,--interactive
.It Fl i,--interactive
ask for each secret whether to re-encrypt it or not
.It Fl k,--keep
do not re-encrypt any secret
.It Fl p,--path
operate on the
.Em recipient
list in the given subfolder instead of the root of the password store
.El
.Ss insert
.Nm
.Cm insert
.Op Fl e,--echo | Fl m,--multiline
.Op Fl f,--force
.Ar pass-name
.Ar ...
.Pp
This subcommand adds new secrets in the database, using the provided data
from the standard input.
By default asks before overwriting an existing secret, and it reads a single
secret line after turning off the console echo, and reads it a second time for
confirmation.
.Pp
The options are as follows:
.Bl -tag -compact -width \-m,--multiline
.It Fl e,--echo
read a single line once without manipulating the standard input
.It Fl m,--multiline
read an arbitrary amount of lines from the standard input, without trying to
manipulate the console, until the end of input or a blank line is entered
.It Fl f,--force
overwrite an existing secret without asking
.El
.Ss list
.Nm
.Op Cm list
.Op r,--raw
.Op Ar subfolder ...
.Pp
This subcommand displays the given subfolders as a tree or as a raw list, or
the whole store when no subfolder is specified.
.Pp
Note that when a secret is given instead of a subfolder, the
.Cm show
command will be used instead, without any warning or error.
.Pp
The options are as follows:
.Bl -tag -compact -width \-r,--raw
.It Fl r,--raw
display the results as a raw list of secrets.
.El
.Ss move
.Nm
.Cm move
.Op Fl e,--reencrypt | Fl i,--interactive | Fl k,--keep
.Op Fl f,--force
.Ar old-path ... new-path
.Pp
This subcommand moves or renames secrets and subfolders recursively,
using the same positional argument scheme as
.Xr mv 1 .
By default it asks before overwriting an existing secret and it re-encrypts
the secret when the destination has a different
.Em recipient
list.
.Pp
The options are as follows:
.Bl -tag -compact -width \-i,--interactive
.It Fl e,--reencrypt
always re-encrypt secrets
.It Fl f,--force
overwrite existing secrets without asking
.It Fl i,--interactive
asks whether to re-encrypt or not for each secret
.It Fl k,--keep
never re-encrypt secrets
.El
.Ss random
.Nm
.Cm random
.Op Ar pass-length Op Ar character-set
.Pp
This subcommand generates a new secret, like the
.Cm generate
subcommand, then directly displays on the standard output without storing it.
.Ss reencrypt
.Nm
.Cm reencrypt
.Op Fl d,--deep
.Op Fl i,--interactive
.Ar pass-name|subfolder
.Ar ...
.Pp
This subcommand re-encrypts in place the given secrets, and all the secrets
recursively in the given subfolders.
.Pp
The options are as follows:
.Bl -tag -compact -width \-i,--interactive
.It Fl d,--deep
re-encrypt subfolders with their own recipient list
.It Fl i,--interactive
asks whether to re-encrypt or not for each secret
.El
.Ss show
.Nm
.Op Cm show
.Oo
.Fl c,--clip Op Ar line-number
|
.Fl q,--qrcode Op Ar line-number
.Oc
.Ar pass-name
.Ar ...
.Pp
This subcommand decrypts the given secrets and by default displays the
whole text on the standard output.
.Pp
Note that when a subfolder is given instead of a secret, the
.Cm list
command will be used instead, without any warning or error.
.Pp
The options are as follows:
.Bl -tag -compact -width \-q,--qrcode
.It Fl c,--clip
paste the given line (by default the first line) of the secret into the
clipboard instead of using the standard output
.It Fl q,--qrcode
display the given line (by default the first line) of the secret as a QR-code
instead of using the standard output
.El
.Ss version
.Nm
.Cm version
.Pp
This subcommand displays on the standard output the version and author
list.
.Sh ENVIRONMENT
The following environment variables affect the execution of
.Nm :
.\" The largest symbol is actually PASSWORD_STORE_CHARACTER_SET_NO_SYMBOLS
.\" but that compresses the second columns way too much.
.Bl -tag -width XXXXXXXX
.It Ev CLICOLOR_FORCE
when set to a non-empty value, and
.Ev NO_COLOR
is unset or empty, use ANSI escape sequences to color the output even when not
writing to a terminal
.It Ev EDITOR
editor command to use instead of
.Xr vi 1
when
.Ev VISUAL
is not set
.It Ev LC_CTYPE
when it contains
.Qq UTF ,
the tree is displayed using Unicode graphic characters instead of ASCII
.It NO_COLOR
when set to a non-empty value, no ANSI escape sequence is output
.It Ev PASHAGE_AGE
external command to use instead of
.Xr age 1
.It Ev PASHAGE_DIR
database directory to use instead of
.Pa ~/.passage/store
.It Ev PASHAGE_IDENTITIES_FILE
.Em identity
file to use instead of
.Pa ~/.passage/identities
.It Ev PASSAGE_AGE
external command to use instead of
.Xr age 1
when
.Ev PASHAGE_AGE
is unset
.It Ev PASSAGE_DIR
database directory to use instead of
.Pa ~/.passage/store
when
.Ev PASHAGE_DIR
is unset
.It Ev PASSAGE_IDENTITIES_FILE
.Em identity
file to use instead of
.Pa ~/.passage/identities
when
.Ev PASHAGE_IDENTITIES_FILE
is unset
.It Ev PASSWORD_STORE_CHARACTER_SET_NO_SYMBOLS
default character set to use with
.Xr tr 1
when
.Fl n
is specified, instead of
.Qq [:alnum:]
.It Ev PASSWORD_STORE_CHARACTER_SET
character set to use with
.Xr tr 1
when no character set requirement is specified for the
.Cm generate
or
.Cm random
commands, instead of
.Qq [:punct:][:alnum:]
.It Ev PASSWORD_STORE_CLIP_TIME
number of second before clearing the clipboard when
.Fl c
is used, instead of 45
.It Ev PASSWORD_STORE_DIR
database directory to use instead of
.Pa ~/.passage/store
when both
.Ev PASHAGE_DIR
and
.Ev PASSAGE_DIR
are unset
.It Ev PASSWORD_STORE_GENERATED_LENGTH
number of characters in the generated secret when not explicitly given,
instead of 25
.It Ev PASSWORD_STORE_X_SELECTION
selection to use when
.Fl c
and
.Xr xclip 1
are used, instead of
.Qq clipboard
.It Ev TMPDIR
temporary directory for the decrypted file to use instead of
.Pa /tmp
when
.Pa /dev/shm
is not available
.It Ev VISUAL
editor command to use instead of
.Xr vi 1
.El
.Sh FILES
Most paths used by
.Nm
can be configured through environment variables.
Here are the defaults:
.Bl -tag -width XXXXXXXX
.It Pa ~/.passage/identities
.Xr age 1
.Em identities
to use for decryption
.It Pa ~/.passage/store/
Root directory of the password store
.It Pa /dev/shm/
Temporary directory
.Pq when available
.It Pa /dev/urandom
Source of entropy
.It Pa /tmp
Temporary directory
.Po
when
.Pa /dev/sdm
is not available
.Pc
.El
.Sh SEE ALSO
.Xr age 1 ,
.Xr git 1 ,
.Xr tr 1 ,
.Xr vi 1 ,
.Xr xclip 1 .
.Sh AUTHORS
.Nm
was written by
.An Natasha Kerensikova ,
based on earlier work by:
.An Jason A. Donenfeld
.Pq password-store
.An Filippo Valsorda
.Pq passage
.An Dylan Araps
.Pq pash


completions/pashage.bash
# completion file for bash

# Copyright (C) 2025  Natasha Kerensikova, based on password-store completion:
# Copyright (C) 2012 - 2014 Jason A. Donenfeld <Jason@zx2c4.com> and
# Brian Mattern <rephorm@rephorm.com>. All Rights Reserved.
# This file is licensed under the GPLv2+. Please see COPYING for more information.

_pashage_complete_entries () {
	local prefix="${PASHAGE_DIR:-${PASSAGE_DIR:-${PASSWORD_STORE_DIR:-${HOME}/.passage/store}}}"
	prefix="${prefix%/}/"
	local suffix=".age"
	local autoexpand=${1:-0}

	local IFS=$'\n'
	local items=($(compgen -f $prefix$cur))

	# Remember the value of the first item, to see if it is a directory. If
	# it is a directory, then don't add a space to the completion
	local firstitem=""
	# Use counter, can't use ${#items[@]} as we skip hidden directories
	local i=0 item

	for item in ${items[@]}; do
		[[ $item =~ /\.[^/]*$ ]] && continue

		# if there is a unique match, and it is a directory with one entry
		# autocomplete the subentry as well (recursively)
		if [[ ${#items[@]} -eq 1 && $autoexpand -eq 1 ]]; then
			while [[ -d $item ]]; do
				local subitems=($(compgen -f "$item/"))
				local filtereditems=( ) item2
				for item2 in "${subitems[@]}"; do
					[[ $item2 =~ /\.[^/]*$ ]] && continue
					filtereditems+=( "$item2" )
				done
				if [[ ${#filtereditems[@]} -eq 1 ]]; then
					item="${filtereditems[0]}"
				else
					break
				fi
			done
		fi

		# append / to directories
		[[ -d $item ]] && item="$item/"

		item="${item%$suffix}"
		COMPREPLY+=("${item#$prefix}")
		if [[ $i -eq 0 ]]; then
			firstitem=$item
		fi
		let i+=1
	done

	# The only time we want to add a space to the end is if there is only
	# one match, and it is not a directory
	if [[ $i -gt 1 || ( $i -eq 1 && -d $firstitem ) ]]; then
		compopt -o nospace
	fi
}

_pashage_complete_folders () {
	local prefix="${PASHAGE_DIR:-${PASSAGE_DIR:-${PASSWORD_STORE_DIR:-${HOME}/.passage/store}}}"
	prefix="${prefix%/}/"

	local IFS=$'\n'
	local items=($(compgen -d $prefix$cur))
	for item in ${items[@]}; do
		[[ $item == $prefix.* ]] && continue
		COMPREPLY+=("${item#$prefix}/")
	done
}

_pashage()
{
	COMPREPLY=()
	local cur="${COMP_WORDS[COMP_CWORD]}"
	local commands="copy cp delete edit find generate git gitconfig grep help init insert list ls move mv random re-encrypt reencrypt remove rm show version"
	if [[ $COMP_CWORD -gt 1 ]]; then
		local lastarg="${COMP_WORDS[$COMP_CWORD-1]}"
		case "${COMP_WORDS[1]}" in
			init)
				if [[ $lastarg == "-p" || $lastarg == "--path" ]]; then
					_pashage_complete_folders
					compopt -o nospace
				else
					COMPREPLY+=($(compgen -W "-i --interactive -k --keep -p --path" -- ${cur}))
				fi
				;;
			ls|list|edit)
				_pashage_complete_entries
				;;
			show|-*)
				COMPREPLY+=($(compgen -W "-c --clip -q --qrcode" -- ${cur}))
				_pashage_complete_entries 1
				;;
			insert)
				COMPREPLY+=($(compgen -W "-e --echo -m --multiline -f --force" -- ${cur}))
				_pashage_complete_entries
				;;
			generate)
				COMPREPLY+=($(compgen -W "-n --no-symbols -c --clip -f --force -i --in-place -m --multiline -q --qrcode -t --try" -- ${cur}))
				_pashage_complete_entries
				;;
			cp|copy|move|mv|rename)
				COMPREPLY+=($(compgen -W "-f --force -e --reencrypt -i --interactive -k --keep" -- ${cur}))
				_pashage_complete_entries
				;;
			rm|remove|delete)
				COMPREPLY+=($(compgen -W "-r --recursive -f --force" -- ${cur}))
				_pashage_complete_entries
				;;
			re-encrypt|reencrypt)
				COMPREPLY+=($(compgen -W "-d --deep -i --interactive" -- ${cur}))
				_pashage_complete_entries
				;;
			git)
				COMPREPLY+=($(compgen -W "init push pull config log reflog rebase" -- ${cur}))
				;;
		esac
	else
		COMPREPLY+=($(compgen -W "${commands}" -- ${cur}))
		_pashage_complete_entries 1
	fi
}

complete -o filenames -F _pashage pashage
